lib/statsd/instrument/compiled_metric.rb

```diff
@@ -0,0 +1,287 @@
+    #
+    # Example:
+    #   CheckoutMetric = StatsD::Instrument::CompiledMetric::Counter.define(
+    #     name: "checkout.completed",
+    #     static_tags: { service: "web" },
+    #     tags: { shop_id: Integer, user_id: Integer }
+    #   )
+    #
+    #   # Later, emit with minimal allocations:
+    #   CheckoutMetric.increment(shop_id: 123, user_id: 456, value: 1)
+    class CompiledMetric
+      # Default maximum number of unique tag combinations to cache before clearing
+      # the cache to prevent unbounded memory growth
+      DEFAULT_MAX_TAG_COMBINATION_CACHE_SIZE = 5000
+
+      class << self
+        # Defines a new compiled metric class with the given configuration.
+        #
+        # @param name [String] The metric name
+        # @param static_tags [Hash{Symbol, String => String, Integer, Float}] Tags with fixed values
+        # @param tags [Hash{Symbol, String => Class}] Tags with dynamic values (Integer, Float, or String)
+        # @param no_prefix [Boolean] If true, skip the StatsD prefix and default_tags
+        # @param max_cache_size [Integer] Maximum tag combinations to cache before clearing
+        # @return [Class] A new CompiledMetric subclass configured for this metric
+        def define(name:, static_tags: {}, tags: {}, no_prefix: false, max_cache_size: DEFAULT_MAX_TAG_COMBINATION_CACHE_SIZE)
+          client = StatsD.singleton_client
+
+          # Build the tag template string
+          tags_str =
+            tags.map do |(k, v)|
+              tag_name = normalize_tag_name(k)
+              placeholder =
+                if v == String
+                  "%s"
+                elsif v == Integer
+                  "%d"
+                elsif v == Float
+                  "%f"
+                else
+                  raise ArgumentError, "Unsupported tag value type: #{v}. Use String, Integer, or Float class."
+                end
+              "#{tag_name}:#{placeholder}"
+            end
+
+          static_tags_str =
+            static_tags.map do |(k, v)|
+              tag_name = normalize_tag_name(k)
+              value = normalize_tag_value(v)
+              "#{tag_name}:#{value}"
+            end
+
+          # Add default_tags from client unless no_prefix is true
+          default_tags_str = []
+          unless no_prefix
+            default_tags_str = (client.default_tags || []).map do |tag|
+              normalize_tag_value(tag)
+            end
+          end
+
+          all_tags = (default_tags_str + static_tags_str + tags_str).join(",")
+
+          # Create a new class for this specific metric
+          # Using classes instead of instances for better YJIT optimization
+          metric_class = Class.new(self) do
+            @name = normalize_name(name)
+            @type = type
+
+            # Build prefix: only add it if no_prefix is false AND a prefix exists
+            @prefix = ""
+            if !no_prefix && client.prefix
+              @prefix = client.prefix + "_"
```

Comment:
This is not correct, prefix is usually "glued" together with metric name with `.` which is default separator in StatsD. Lets make sure this reflects on tests as well.